Add new third party settings needed by publish

dsenalik opened the pull request:
# New Feature

### Issue #1924

### ~~Depends on #2001~~ (merged)

### Closes #1994

### Tripal Version: 4.x

## Description
This PR adds two new third party settings intended for use by the publish service.
These two settings are used for publishing content types where more than one bundle is based on a single chado base table. For example, both the "gene" and "mRNA" content types are derived from the `feature` table.
The third party settings allow publish to know where to find the term that defines the bundle, whether in the base table, or in a property table.


## Testing?
This PR isolates the described changes from other changes in the publish PR #1991 to simplify review.

Testing involves:
1. Build a docker on 4.x branch
2. Import all of the type collections
3. Switch to this branch and test the update hook `drush updatedb`
4. Verify that the settings have now been added
```
$bundles = ['genetic_map', 'QTL', 'sequence_variant', 'genetic_marker', 'phenotypic_marker',
            'gene', 'mrna', 'physical_map', 'genome_assembly', 'genome_annotation', 'genome_project',
            'germplasm', 'breeding_cross', 'germplasm_variety', 'ril'];
foreach ($bundles as $bundle) {
  $entity_type_manager = \Drupal::entityTypeManager();
  $entity_type = $entity_type_manager->getStorage('tripal_entity_type')->load($bundle);
  $entity_type_chado_base_table = $entity_type->getThirdPartySetting('tripal', 'chado_base_table');
  $entity_type_bundle_type_table = $entity_type->getThirdPartySetting('tripal', 'bundle_type_table');
  $entity_type_bundle_type_column = $entity_type->getThirdPartySetting('tripal', 'bundle_type_column');
  print "bundle=\"$bundle\"\n";
  print "  base_table=\"$entity_type_chado_base_table\"\n";
  print "  type_table=\"$entity_type_bundle_type_table\"\n";
  print "  type_column=\"$entity_type_bundle_type_column\"\n";
}
```
expected output
```
bundle="genetic_map"
  base_table="featuremap"
  type_table="featuremapprop"
  type_column="type_id"
bundle="QTL"
  base_table="feature"
  type_table="feature"
  type_column="type_id"
bundle="sequence_variant"
  base_table="feature"
  type_table="feature"
  type_column="type_id"
bundle="genetic_marker"
  base_table="feature"
  type_table="feature"
  type_column="type_id"
bundle="phenotypic_marker"
  base_table="feature"
  type_table="feature"
  type_column="type_id"
bundle="gene"
  base_table="feature"
  type_table="feature"
  type_column="type_id"
bundle="mrna"
  base_table="feature"
  type_table="feature"
  type_column="type_id"
bundle="physical_map"
  base_table="featuremap"
  type_table="featuremapprop"
  type_column="type_id"
bundle="genome_assembly"
  base_table="analysis"
  type_table="analysisprop"
  type_column="type_id"
bundle="genome_annotation"
  base_table="analysis"
  type_table="analysisprop"
  type_column="type_id"
bundle="genome_project"
  base_table="project"
  type_table="projectprop"
  type_column="type_id"
bundle="germplasm"
  base_table="stock"
  type_table="stock"
  type_column="type_id"
bundle="breeding_cross"
  base_table="stock"
  type_table="stock"
  type_column="type_id"
bundle="germplasm_variety"
  base_table="stock"
  type_table="stock"
  type_column="type_id"
bundle="ril"
  base_table="stock"
  type_table="stock"
  type_column="type_id"
```
5. And for good measure, a docker built on this branch should already include these settings.

### Test creating a new content type - adding type field manually
1. Creating a new content type. Tripal -> Page Structure -> "+ Add custom type"
2. I created a "Tripal Project" content type based on the project table.
3. Add a "Tripal Project Name" field using the "Chado String Field Type" field - this gets the base table populated.
4. Add a "Tripal Project Type" field using the "Chado Type Reference" field. Term should match the bundle term. You will want to check the new box "This term defines the bundle"
![issue1924_gui-term-defines-bundle](https://github.com/user-attachments/assets/54696450-d24b-4cb3-bc18-2972b3c9e337)

### Test creating a new content type - adding type field manually
1-3. Repeat the steps 1-3 above, but then
4. Discover new fields
5. Edit the type field and you can then check the  "This term defines the bundle" box and save.

dsenalik committed:
add new third party settings

laceysanderson committed:
Merge branch '4.x' into tv4g1-issue1924-third-party-settings

laceysanderson reviewed (commented):
I started a fresh docker on this branch then
✅ confirmed that the settings were set properly for all the existing general content types
✅ imported all the other content type collections and confirmed they're settings were correct as well.
❌ Tried to create a new custom content type to check it but got this validation error since type is in the default URL pattern but a new content type doesn't have any fields yet 🙈 This is not related to this PR but must have slipped in elsewhere.
![image](https://github.com/user-attachments/assets/f14effca-6b01-4729-be79-3f5216a113ce)
❌ Add a type field and have the 3rd party settings added.
Your instructions may not have been clear enough for me to get this working right 😞 Since it is really finicky we should look at making the type field self discoverable so people who create custom bundles can use the check fields to have this set properly.

laceysanderson commented on tripal_chado/src/TripalField/ChadoFieldItemBase.php:
```diff
@@ -411,6 +411,22 @@ public static function storageSettingsFormSubmitBaseTable(array $form, FormState
       $entity_type->setThirdPartySetting('tripal', 'chado_base_table', $base_table);
       $entity_type->save();
     }
+
+    // If this is the additional type field, it is used to specify a subset of
+    // a chado table for this bundle, e.g. for feature table it could be 'gene'.
+    // The first time this field is entered, store third party settings in the
+    // entity for the table and column where this term will be stored, so that
+    // publish will be able to restrict by the term.
+    $type_fkey = $settings['storage_plugin_settings']['type_fkey'] ?? NULL;
+    if ($type_fkey) {
+      $type_table = $entity_type->getThirdPartySetting('tripal', 'chado_type_table');
+      if (!$type_table) {
+        list($type_table, $type_column) = explode(self::$table_column_delimiter, $type_fkey, 2);
+        $entity_type->setThirdPartySetting('tripal', 'chado_type_table', $type_table);
+        $entity_type->setThirdPartySetting('tripal', 'chado_type_column', $type_column);
+        $entity_type->save();
+      }
+    }
```
I feel like this should be in the additional type field. Then it doesn't have to do this check. That field can just override this method, call the parent version first and then do this.

dsenalik replied:
This has been moved

dsenalik replied:
And the type field is now discoverable

laceysanderson commented on tripal_chado/tripal_chado.install:
```diff
@@ -575,3 +575,44 @@ function tripal_chado_update_10406() {
+ * content types derived from one Chado base table. PR#1991
+ */
+function tripal_chado_update_10407() {
+  $settings = [
+    // analysis table
+    'genome_assembly' => 'analysisprop',
+    'genome_annotation' => 'analysisprop',
+    // feature table
+    'gene' => 'feature',
+    'mrna' => 'feature',
+    'QTL' => 'feature',
+    'sequence_variant' => 'feature',
+    'genetic_marker' => 'feature',
+    'phenotypic_marker' => 'feature',
+    // featuremap table
+    'genetic_map' => 'featuremapprop',
+    'physical_map' => 'featuremapprop',
+    // project table
+    'genome_project' => 'projectprop',
+    // stock table
+    'germplasm' => 'stock',
+    'breeding_cross' => 'stock',
+    'germplasm_variety' => 'stock',
+    'ril' => 'stock',
+  ];
```
What about Tripal 4 sites that have their own custom content types already? They will not get touched by this update hook.

dsenalik replied:
> What about Tripal 4 sites that have their own custom content types already? They will not get touched by this update hook.

The update hook is now generic - it uses the "fixed_type" settings.

laceysanderson replied:
Confirmed resolved in new update hook.

laceysanderson reviewed (commented):
I worry there is no way for the 3rd party setting to be added for a content type created through YAML by an extension module who doesn't know to provide that info in the yaml file 🤔

dsenalik commented:
I forgot to mention that validation problem in my testing instructions 🙊 - I made an issue for it #1996
I will look at your other comments...

dsenalik committed:
make additional type field discoverable

dsenalik committed:
detect existing field by CV term rather than by name

dsenalik committed:
Merge branch '4.x' into tv4g1-issue1924-third-party-settings

dsenalik committed:
helper function to reduce complexity

dsenalik committed:
Merge branch 'tv4g0-issue2000-fix-failing-docker-build-on-drush-13.3.0' into tv4g1-issue1924-third-party-settings

dsenalik committed:
Merge branch '4.x' into tv4g1-issue1924-third-party-settings

dsenalik committed:
remove chado from setting name

dsenalik committed:
fixed_type from additional type saved to entity

dsenalik committed:
rework update hook 10407 to be generic

dsenalik committed:
store fixed_value

dsenalik committed:
Finish update hook

dsenalik committed:
Merge branch '4.x' into tv4g1-issue1924-third-party-settings

dsenalik committed:
quick fix for issue #1994

dsenalik committed:
typo fix

dsenalik commented:
> I worry there is no way for the 3rd party setting to be added for a content type created through YAML by an extension module who doesn't know to provide that info in the yaml file 🤔

This can now be done by adding or editing the additional type field.

laceysanderson commented on tripal_chado/src/Plugin/Field/FieldType/ChadoAdditionalTypeTypeDefault.php:
```diff
@@ -394,4 +465,99 @@ public function isCompatible(TripalEntityType $entity_type) : bool {
+      ],
+      'settings' => [
+        'termIdSpace' => $termIdSpace,
+        'termAccession' => $termAccession,
```
The newly discovered field should have the fixed value set by default. This allows that checkbox to be checked by default and makes the type table/column be automatically detected when the type field is added by check fields rather then requiring an edit to the field.

I know there are cases where this field doesn't define the content type but that feels like the exception rather then the default 🤔 and you do specifically mention the fixed value in the field description you generate :wink'

```suggestion
        'termAccession' => $termAccession,
        'fixed_value' => $fixed_value,
```

dsenalik replied:
I removed the default of checked with commit https://github.com/tripal/tripal/pull/1992/commits/0bdb0f8801a19702fb33a4d6ec9b857e085e2147
My thinking was that this was a safer default, however on further consideration I agree with your thinking that it is a sensible default, as those other use cases will be uncommon.

laceysanderson reviewed (commented):
This does now work when adding a custom content type and I like that a simple edit will fix things if your YAML was missing information!

That said, it took me a bit of detective work to find the checkbox I needed to select since it was hidden in the controlled vocabulary term fieldset which is collapsed by default... that said this is an appropriate place to put it 🤷‍♀️ so I think I'm ok with that as long as the discovered field has this checked by default rather then requiring edit. Alternatively, we could add a drupal message letting people know they may want to select this for the type field specifically when it is edited?

dsenalik committed:
Update tripal_chado/src/Plugin/Field/FieldType/ChadoAdditionalTypeTypeDefault.php

I removed this with commit https://github.com/tripal/tripal/pull/1992/commits/0bdb0f8801a19702fb33a4d6ec9b857e085e2147 - I was thinking the safer default was to leave it off, however I think you are correct, this would be a better default.

Co-authored-by: Lacey-Anne Sanderson <lacey.sanderson@usask.ca>

laceysanderson reviewed (approved):
Tested again and it looks good 👍
